Task/58634 revise what happens on work package description edits when two users are concurrently editing

jjabari-op opened the pull request:
[58634](https://community.openproject.org/projects/openproject/work_packages/58634)

jjabari-op committed:
[#58634] Revise what happens on work package description edits when two users are concurrently editing

https://community.openproject.org/work_packages/58634

jjabari-op committed:
check for active edit state and warn about potential conflict/override

jjabari-op committed:
resolving conflict override situations

jjabari-op committed:
Merge branch 'dev' into task/58634-revise-what-happens-on-work-package-description-edits-when-two-users-are-concurrently-editing

jjabari-op committed:
fixing specs

akabiru commented on app/controllers/work_packages_controller.rb:
```diff
@@ -40,6 +41,8 @@ class WorkPackagesController < ApplicationController
                 :check_allowed_export,
                 :protect_from_unauthorized_export, only: %i[index export_dialog]
+
+  before_action :authorize, only: :show_conflict_flash_message
```
🟠  Q: What's the reason behind authorizing only on show conflict? I would presume all controller actions here are to be authorized?

oliverguenther replied:
This is fine, as `authorize_on_work_package` exists

akabiru commented on app/components/work_packages/update_conflict_component.rb:
```diff
@@ -0,0 +1,72 @@
+      end
+    end
+
+    def call
```
🟢 I reckon using call is fine here given the footprint but I should mention we agreed conventionally to always opt for a template.

akabiru commented on frontend/src/app/features/work-packages/services/notifications/work-package-notification.service.ts:
```diff
@@ -32,14 +32,21 @@ import { HalResourceNotificationService } from 'core-app/features/hal/services/h
 import { HalResource } from 'core-app/features/hal/resources/hal-resource';
-
+import { TurboRequestsService } from 'core-app/core/turbo/turbo-requests.service';
+import { ConfigurationService } from 'core-app/core/config/configuration.service';
```
eslint error 🧹

akabiru commented on frontend/src/app/features/work-packages/services/notifications/work-package-notification.service.ts:
```diff
@@ -53,14 +60,26 @@ export class WorkPackageNotificationService extends HalResourceNotificationServi
-        },
-      });
+      if (this.primerizedActivitiesEnabled) {
+        // currently we do not have a programmatic way to show the primer flash messages
+        // so we just do a request to the server to show it
+        // should be refactored once we have a programmatic way to show the primer flash messages!
```
🟠 Q: Would you imagine sth similar to the `ToastService`? I reckon with the hotwire approach, it's fine to have the template server rendered either way.

oliverguenther replied:
My expectation is that you would pass this state to the backend request for updating the activity, which returns a toast in case of a conflict. I briefly talked to Jonas about this and agreed that a separate request is an acceptable hack for now

akabiru commented on spec/features/activities/work_package/activities_spec.rb:
```diff
@@ -980,18 +980,224 @@
       # otherwise the update is already picked up by the initial (async) workpackage attributes update called in the connect hook
       # and we wouldn't test the polling based update below
-      sleep 2
+      sleep 2 # rubocop:disable OpenProject/NoSleepInFeatureSpecs
```
🔴  we since added [rspec-wait](https://github.com/laserlemon/rspec-wait) in https://github.com/opf/openproject/pull/16761 and can already leverage `wait_for` in place of static sleep times.

Ex. https://github.com/opf/openproject/blob/dev/spec/features/activities/work_package/activities_spec.rb#L373

oliverguenther replied:
Indeed, do not wait for 2 seconds. You didn't introduce it here but I'm certain you can rewrite the expectation quickly to avoid this line (and the copied one below)

akabiru commented on spec/features/activities/work_package/activities_spec.rb:
```diff
@@ -980,18 +980,224 @@
+      # we need to wait a bit before triggering the update below
+      # otherwise the update is already picked up by the initial (async) workpackage attributes update called in the connect hook
+      # and we wouldn't test the polling based update below
+      sleep 2 # rubocop:disable OpenProject/NoSleepInFeatureSpecs
```
🔴 ditto

akabiru commented on spec/features/activities/work_package/activities_spec.rb:
```diff
@@ -980,18 +980,224 @@
+    it "shows the updated work package attribute without reload after switching back to the activity tab", :aggregate_failures do
+      # wait for the latest comments to be loaded before proceeding!
+      activity_tab.expect_journal_notes(text: "First comment by member")
+      wp_page.expect_attributes(subject: work_package.subject)
```
_Edit: I see it's a simulated delay, hence `wait_for` would not exactly solve the issue._ 

---

🟠 Q: Seems we're calling `wp_page.expect_attributes(subject: work_package.subject)` twice - can we replace it with a single wait_for statement?

```
wait_for { page }.to have_css(".inline-edit--container.subject", text: work_package.subject)
```

perhaps in this case, inline assertions are more concise

akabiru commented on spec/features/activities/work_package/activities_spec.rb:
```diff
@@ -980,18 +980,224 @@
+
+    it "raises a conflict warning when the work package is updated by another user while the current user is editing",
+       :aggregate_failures do
+      using_session(:admin) do
```
Nice, I didn't know `using_session` was a thing 👍🏾

akabiru reviewed (changes requested):
Looking good and manual QA successful! Just a few remarks and failing build 👍🏾 

---

Perhaps  @oliverguenther could have a quick look too?

akabiru commented:
⚠️ ⚠️ Remember to rebase onto `release/15.0` ⚠️ ⚠️

akabiru committed:
tests[Op#58634]: use static string to avoid stale comparisons

Use static strings instead of `work_package.subject` to simulate previous subject value and after update. Reading the subject
from the work_package will generally always return the updated value- which is not the desired case.

oliverguenther reviewed (commented):
Jonas and I aligned on the changes in the PR already, I did not find additional remarks from what Kabiru already mentioned there

akabiru commented:
@jjabari-op please assist with the merge conflicts- I can then rebase onto release/15.0

jjabari-op commented:
Continue in #17152 which cherry-picked the relevant commits in order to cleanly target the release branch
